# Analysis of Lenovo K5 Note

In [1]:
import re

In [2]:
file = open(r'F:/lenovo.txt', 'r')

In [3]:
original_text = file.read()

### Splitting the raw text on the newline character

In [4]:
review_list = original_text.split('\n')

In [5]:
print(review_list)

['Lenovo K5 Note', '4?', 'Great device ------ go for it.....', 'I wont talk much about the specifications as it is already clearly mentioned on flipkart. I will share only my experience with this device.', '', 'I am using the phone from last 1 and half month. Will break down my review in 3 parts. The pros, cons and what more could be done.', "First let's see the pros that make it the best value for money phone currently in the market.", '', '* Battery : easily getting around 6 hours of screen on time. Battery is an issue with smartphones today. This phone is just a beas...', 'READ MORE', 'SURENDRA KUMAR MEENA', '', 'Certified Buyer', '28 Oct, 2016', '9216', '1431', '5?', 'Simply awesome', 'AMAZING PHONE ', 'DISPLAY-10/10 display is awesome', 'CAMERA-10/10 i think better than redmi note 4', 'BATTERY-8/10 good battery last for 1 day', 'PROCESSOR-10/8 not fast as redmi note 4 ', 'SOUND-10/10 awesome dolby digital sound', '', 'AMAZING PHONE ', 'LENOVO', 'shivam sharma', '', 'Certified Buye

In [6]:
review_list = review_list[1:]

In [7]:
print(len(review_list))

1353


### Removing empty strings from the list

In [8]:
for i in review_list:
    if(i == ''):
        review_list.remove(i)

In [9]:
print(len(review_list))

1190


### Building a list with individual reviews as its elements

In [10]:
individual_reviews = []
temp = ''
c = 0
for i in review_list:
    if(re.match(r'[\d]\?', i)):
        individual_reviews.append(temp)
        temp = '' + i
    elif(i == 'Certified Buyer'):
        c+=1
    else:
        temp = temp + ' ' + i

In [11]:
print(individual_reviews)

['', "4? Great device ------ go for it..... I wont talk much about the specifications as it is already clearly mentioned on flipkart. I will share only my experience with this device. I am using the phone from last 1 and half month. Will break down my review in 3 parts. The pros, cons and what more could be done. First let's see the pros that make it the best value for money phone currently in the market. * Battery : easily getting around 6 hours of screen on time. Battery is an issue with smartphones today. This phone is just a beas... READ MORE SURENDRA KUMAR MEENA 28 Oct, 2016 9216 1431", '5? Simply awesome AMAZING PHONE  DISPLAY-10/10 display is awesome CAMERA-10/10 i think better than redmi note 4 BATTERY-8/10 good battery last for 1 day PROCESSOR-10/8 not fast as redmi note 4  SOUND-10/10 awesome dolby digital sound AMAZING PHONE  LENOVO shivam sharma 18 Mar, 2017 1530 333', "4? Good choice Good chinese Product. My comparison with redmi note 3 and some drawbacks i saw in lenovo a

In [12]:
individual_reviews = individual_reviews[1:]

In [13]:
len(individual_reviews)

119

### All are certified customers

In [14]:
c

119

In [15]:
individual_reviews = []
date = []
temp = ''
for i in review_list:
    if(re.match(r'[\d]\?', i)):
        individual_reviews.append(temp)
        temp = '' + i
    elif(re.match(r'\d{1,2} (jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)', i.lower())):
        date.append(i.lower())
    elif(i == 'Certified Buyer'):
        continue
    elif(re.match(r'\d{1,2}(\. | [a-z]+)', i)):
        temp = temp + ' ' + i.lower()
    elif(re.match(r'\d{1,5}', i)):
        continue
    else:
        temp = temp + ' ' + i.lower()

In [16]:
print(individual_reviews[1])

4? great device ------ go for it..... i wont talk much about the specifications as it is already clearly mentioned on flipkart. i will share only my experience with this device. i am using the phone from last 1 and half month. will break down my review in 3 parts. the pros, cons and what more could be done. first let's see the pros that make it the best value for money phone currently in the market. * battery : easily getting around 6 hours of screen on time. battery is an issue with smartphones today. this phone is just a beas... read more surendra kumar meena


In [17]:
individual_reviews = individual_reviews[1:]

In [18]:
print(len(date))

119


### Creating separate lists for ratings and reviews

In [19]:
ratings = []
reviews = []
for i in individual_reviews:
    ratings.append(i[0])
    reviews.append(i[3:])

In [20]:
print(len(ratings))

119


In [24]:
print(len(reviews))

119


### Removing stopwords and stemming the remaining words

In [25]:
from nltk.corpus import stopwords

In [26]:
from nltk.stem.porter import PorterStemmer

In [27]:
ps = PorterStemmer()

In [28]:
stem = []
for i in reviews:
    i = i.lower()
    alpha = re.sub('[^a-zA-Z]', ' ', i)
    alpha_list = alpha.split()
    temp = []
    sw = set(stopwords.words('english'))
    for j in alpha_list:
        if(j not in sw):
            temp.append(ps.stem(j))
    string = " ".join(temp)
    stem.append(string)

In [29]:
print(stem)

['great devic go wont talk much specif alreadi clearli mention flipkart share experi devic use phone last half month break review part pro con could done first let see pro make best valu money phone current market batteri easili get around hour screen time batteri issu smartphon today phone bea read surendra kumar meena', 'simpli awesom amaz phone display display awesom camera think better redmi note batteri good batteri last day processor fast redmi note sound awesom dolbi digit sound amaz phone lenovo shivam sharma', 'good choic good chines product comparison redmi note drawback saw lenovo ir blaster redmi note lenovo k video playabl redmi note lenovo batteri mah redmi note mah processor speed good redmi note compar k notif led light blink red blue k redmi note blink color download speed good redmi note use jio sim android process com pop read flipkart custom', 'wonder look guy brought phone day go buy phone let see need g mobil volt true check batteri pretti good heavi user camera g

In [30]:
print(len(stem))

119


In [31]:
'''s = set()
for i in stem:
    l = []
    l = i.split()
    s.update(l)'''

's = set()\nfor i in stem:\n    l = []\n    l = i.split()\n    s.update(l)'

In [32]:
#print(len(s))

In [33]:
temp = " ".join(stem) 

In [34]:
print(temp)

great devic go wont talk much specif alreadi clearli mention flipkart share experi devic use phone last half month break review part pro con could done first let see pro make best valu money phone current market batteri easili get around hour screen time batteri issu smartphon today phone bea read surendra kumar meena simpli awesom amaz phone display display awesom camera think better redmi note batteri good batteri last day processor fast redmi note sound awesom dolbi digit sound amaz phone lenovo shivam sharma good choic good chines product comparison redmi note drawback saw lenovo ir blaster redmi note lenovo k video playabl redmi note lenovo batteri mah redmi note mah processor speed good redmi note compar k notif led light blink red blue k redmi note blink color download speed good redmi note use jio sim android process com pop read flipkart custom wonder look guy brought phone day go buy phone let see need g mobil volt true check batteri pretti good heavi user camera good compar 

## Passing the string 'temp' to pos_tag will not yield the desired result

### Tokenizing the words

In [35]:
from nltk import word_tokenize

In [36]:
tokens = word_tokenize(temp)

In [37]:
print(tokens)

['great', 'devic', 'go', 'wont', 'talk', 'much', 'specif', 'alreadi', 'clearli', 'mention', 'flipkart', 'share', 'experi', 'devic', 'use', 'phone', 'last', 'half', 'month', 'break', 'review', 'part', 'pro', 'con', 'could', 'done', 'first', 'let', 'see', 'pro', 'make', 'best', 'valu', 'money', 'phone', 'current', 'market', 'batteri', 'easili', 'get', 'around', 'hour', 'screen', 'time', 'batteri', 'issu', 'smartphon', 'today', 'phone', 'bea', 'read', 'surendra', 'kumar', 'meena', 'simpli', 'awesom', 'amaz', 'phone', 'display', 'display', 'awesom', 'camera', 'think', 'better', 'redmi', 'note', 'batteri', 'good', 'batteri', 'last', 'day', 'processor', 'fast', 'redmi', 'note', 'sound', 'awesom', 'dolbi', 'digit', 'sound', 'amaz', 'phone', 'lenovo', 'shivam', 'sharma', 'good', 'choic', 'good', 'chines', 'product', 'comparison', 'redmi', 'note', 'drawback', 'saw', 'lenovo', 'ir', 'blaster', 'redmi', 'note', 'lenovo', 'k', 'video', 'playabl', 'redmi', 'note', 'lenovo', 'batteri', 'mah', 'redmi

### Parts of Speech tagging

In [38]:
from nltk import pos_tag

In [39]:
tags = pos_tag(tokens)

In [40]:
for i in tags:
    print(i[0], "  ", i[1])

great    JJ
devic    NN
go    VBP
wont    JJ
talk    NN
much    JJ
specif    NN
alreadi    NN
clearli    NN
mention    NN
flipkart    NN
share    NN
experi    VBZ
devic    JJ
use    NN
phone    NN
last    JJ
half    JJ
month    NN
break    NN
review    NN
part    NN
pro    JJ
con    NN
could    MD
done    VBN
first    JJ
let    NN
see    VB
pro    JJ
make    VB
best    JJS
valu    NN
money    NN
phone    NN
current    JJ
market    NN
batteri    NN
easili    VBP
get    VB
around    IN
hour    NN
screen    NN
time    NN
batteri    NN
issu    NN
smartphon    NN
today    NN
phone    NN
bea    NN
read    VBD
surendra    JJ
kumar    FW
meena    FW
simpli    FW
awesom    FW
amaz    FW
phone    NN
display    NN
display    NN
awesom    VBP
camera    NN
think    VBP
better    JJR
redmi    NN
note    NN
batteri    NN
good    JJ
batteri    NN
last    JJ
day    NN
processor    VBD
fast    RB
redmi    JJ
note    NN
sound    NN
awesom    IN
dolbi    NN
digit    NN
sound    NN
amaz    IN
phone    NN
l

kumar    NN
great    JJ
product    NN
super    NN
bala    NN
murugan    JJ
p    JJ
hate    NN
camera    NN
qualitu    NN
poor    JJ
megapixel    NN
camera    NN
megapixel    NN
camera    NN
pratik    NN
saha    VBD
good    JJ
phone    NN
featur    NN
wise    NN
good    JJ
perform    NN
best    JJS
miss    NN
sd    NN
card    NN
slot    NN
gb    NN
inbuilt    JJ
suffici    NN
hybrid    JJ
sim    NN
slot    NN
full    JJ
fill    NN
emerg    VBP
need    NN
galleri    NNS
good    JJ
phone    NN
pramod    NN
tyagi    NN
worst    JJS
mobil    NN
unhappi    JJ
product    NN
bought    VBD
phone    NN
week    NN
back    RB
start    JJ
face    NN
issu    NN
especi    VBP
batteri    NN
drainag    NN
get    NN
also    RB
vinay    VBZ
kadari    JJ
unsatisfactori    JJ
phone    NN
speaker    NN
work    NN
within    IN
day    NN
valu    FW
money    NN
tushar    NN
dey    NN
realli    NN
nice    JJ
batteri    NN
drain    NN
fast    JJ
vinay    NN
thankavelu    NN
utterli    JJ
disappoint    NN
everi  

camera    NN
qualiti    JJ
good    JJ
day    NN
night    NN
see    VBP
nois    JJ
game    NN
exp    NN
simpli    NN
awesom    NN
order    NN
amkett    NN
evo    JJ
pro    FW
control    NN
game    NN
awesom    JJ
flipkart    NN
custom    NN
wow    NN
awesom    NN
tanwang    VBD
khete    JJ
heat    NN
issu    NN
good    JJ
phone    NN
appreanc    NN
heat    NN
issu    VBP
phone    NN
reach    NN
temperatur    NN
degre    NN
play    NN
game    NN
min    NNP
camera    NN
good    JJ
lot    NN
prelod    NNS
app    VBP
amanul    JJ
haqu    NN
got    VBD
defect    JJ
product    NN
flipkart    NN
got    VBD
defect    JJ
product    NN
flipkart    NN
devic    VBZ
get    VB
heat    NN
drain    NN
batteri    IN
fast    JJ
case    NN
first    RB
day    NN
onward    IN
sim    NN
slot    NN
work    NN
request    NN
flipkart    NN
replac    NN
refund    NN
urgent    JJ
basi    NN
ranjeet    NN
choudhari    NN
utterli    JJ
disappoint    NN
excit    NN
phone    NN
main    JJ
featur    NN
like    IN
proc

### Sentiment list from sig_textblob.py

In [41]:
sentiment = [0.26623931623931624, 0.5181818181818181, 0.36999999999999994, 0.37395833333333334, 0.17500000000000002, 0.5733333333333334, 0.3666666666666667, 0.51, 0.5194444444444445, 0.17121212121212126, 0.6466666666666667, 0.39846938775510204, 0.225, 0.5, 0.6599999999999999, 0.9099999999999999, 1.0, 0.57, 0.0, 0.1315340909090909, 0.31533333333333335, 0.7, 0.6499999999999999, 0.85, 0.49026262626262623, 0.9, -0.3499999999999999, 0.33849999999999997, 0.25, -0.22159090909090906, 0.5055555555555556, 0.5866666666666667, 0.75, 1.0, 0.5522222222222222, 0.326923076923077, 0.7666666666666666, 0.85, -0.010000000000000009, 0.24818181818181814, -0.35833333333333334, 0.5333333333333333, 0.2785714285714286, -0.5, 0.6, 0.3513888888888889, 0.72, 0.4733333333333334, -1.0, -0.1287878787878788, 0.2708333333333333, 0.47222222222222227, 0.6166666666666667, -0.76, 0.5642857142857142, -0.445, 0.0, 0.4, -0.75, 0.03545454545454545, 0.722857142857143, -0.140625, -0.19999999999999996, 0.0, 0.10000000000000003, 0.36833333333333335, 0.3, 0.64, 0.46499999999999997, -0.09666666666666668, -0.35, 0.44999999999999996, 0.4785714285714286, 0.33809523809523817, -0.23666666666666658, 0.7, -0.5666666666666668, 0.682, 0.5, 0.2896103896103896, -0.9099999999999998, 0.015000000000000013, 0.7, 0.38541666666666663, 0.7, 0.3770833333333334, 0.37500000000000006, 0.0, 0.5453703703703704, 0.2212121212121212, 0.68, 0.13, 0.29932291666666666, 0.41261904761904766, 0.43427170868347337, 0.39999999999999997, 0.3375, 0.30909090909090914, 0.14666666666666664, 0.26227272727272727, 0.55, 0.3625, -0.2, -0.8, 0.6, 0.75, 0.8, 0.35000000000000003, 0.6233333333333333, -0.12712121212121216, 0.10416666666666666, 0.8571428571428571, 0.5625, 0.3333333333333333, 0.255, 0.1130952380952381, 0.29374999999999996, -0.35, 0.5583333333333333]

### Linear regression

In [42]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

a = pd.DataFrame(sentiment[:99])
b = pd.DataFrame(ratings[:99])

clf = LinearRegression()
clf.fit(a, b)
from textblob import TextBlob
string = input('Enter your review\n')
c = TextBlob(string)
d = c.sentiment.polarity
r = clf.predict(d)
print(int(r))

Enter your review
good phone with poor display
3


### Most frequent words

In [43]:
s = set(tokens)
print(len(s))
l = list(s)

897


In [44]:
word_count = {}
for i in range(897):
    word_count[l[i]] = tokens.count(l[i])

In [45]:
print(word_count)

{'premium': 1, 'alreadi': 1, 'finger': 2, 'sidhdharthsinh': 1, 'happi': 6, 'hang': 10, 'mi': 1, 'super': 6, 'kaustubh': 1, 'ahead': 1, 'fixtur': 1, 'saravanan': 1, 'saha': 1, 'haqu': 1, 'littl': 1, 'brush': 1, 'amkett': 1, 'worth': 8, 'deliveri': 8, 'avail': 4, 'grab': 2, 'weight': 1, 'gorilla': 2, 'ravendra': 1, 'unless': 1, 'par': 1, 'function': 1, 'tiwari': 1, 'featur': 19, 'launch': 1, 'processor': 5, 'divya': 1, 'notch': 1, 'akshay': 1, 'lot': 4, 'effect': 1, 'advantag': 3, 'hybrid': 8, 'doesnt': 1, 'data': 1, 'speaker': 6, 'reach': 1, 'g': 5, 'togath': 1, 'comment': 1, 'purchas': 9, 'advis': 1, 'per': 1, 'think': 8, 'jose': 1, 'mp': 1, 'bettri': 1, 'came': 1, 'combo': 1, 'satisfactori': 1, 'found': 3, 'recommend': 3, 'wont': 2, 'overal': 6, 'anuradha': 1, 'gb': 18, 'show': 1, 'multi': 2, 'sd': 6, 'helio': 2, 'plastic': 1, 'use': 37, 'camera': 39, 'asphalt': 2, 'bea': 1, 'amplifi': 1, 'snapdragon': 1, 'bass': 1, 'ramesh': 1, 'kapadan': 1, 'notif': 1, 'place': 1, 'drawback': 1, 'so

In [46]:
values = list(word_count.values())

In [47]:
keys = list(word_count.keys())

In [48]:
values.index(123)

462

In [49]:
print(values[310])

1


In [50]:
print(keys[310])

vibrat


In [51]:
sum = 0
for i in range(10):
    maxi = max(values)
    i_maxi = values.index(maxi)
    key_maxi = keys[i_maxi]
    sentiment = TextBlob(key_maxi)
    sentiment = sentiment.sentiment.polarity
    print(i + 1, key_maxi, "   ", sentiment)
    sum += sentiment
    values.remove(values[i_maxi])
    keys.remove(keys[i_maxi])

1 phone     0.0
2 good     0.7
3 flipkart     0.0
4 camera     0.0
5 use     0.0
6 product     0.0
7 custom     0.0
8 batteri     0.0
9 lenovo     0.0
10 awesom     0.0


In [52]:
mean_rating = 0
for i in ratings:
    mean_rating += int(i)
mean_rating = mean_rating / len(ratings)

In [53]:
print(mean_rating)

3.7815126050420167


In [54]:
count = 0
sum = 0
for i in individual_reviews:
    if('camera' in i):
        count += 1
        sum += TextBlob(i).sentiment.polarity
print(sum / count)

0.35132355060290116


In [55]:
count = 0
sum = 0
for i in individual_reviews:
    if('battery' in i):
        count += 1
        sum += TextBlob(i).sentiment.polarity
print(sum / count)

0.3506294830243243


In [56]:
count = 0
sum = 0
for i in individual_reviews:
        count += 1
        sum += TextBlob(i).sentiment.polarity
print(sum / count)

0.2996562860442375


## From the given dataset

### The following can be safely infered

* The Overall ratings of the customers for this model is 3.78 which is average.


* All the 119 reviews were made by certified customers making their views more trustworthy.


* The most frequently repeated 10 words from the customer reviews are:

    1 phone
    
    2 good
    
    3 flipkart
    
    4 camera
    
    5 use   
    
    6 product
    
    7 batteri
    
    8 custom 
    
    9 lenovo
    
    10 awesom
    
    All the above words do not have any negative connotations attached to them.
    

* The features of the phone that found the most mentions are 'camera'(0.3513) and 'battery'(0.3506). Predominantly they were made with a positive setting. So the the features that attract the customers more towards the phone are camera and battery. 


* The average sentiment score(mesured with the help of TextBlob package) for the reviews is 0.29(scale: -1 to 1). 


-> From the above it can be safely assumed that eventhough there is nothing so poor about the phone, the problem is there is nothing too tempting for the user too, which is quite evident from the average rating, average sentiment score overall, and for camera and battery. 